### PR TITLE
Descriptor based file I/O

### PR DESCRIPTION
**Pull request by aradi**
It is **really**  ready to be reviewed now. :smile: 

* All file I/O had been converted (apart of a few command line Fortran tools, which do not depend on `libdftbplus`).

* Additionally a fix against obtaining `NaN`s in the timing data had to be implemented as the added test had zero total time, and failed due to `NaN`s in the output.

* Additionally, the NaN-checker output had been improved slightly to make it easier to recognize (I needed quite a while to realize, that the miraculous occasional disappearance of autotest.tag was not a Fortran error, but the NaN-checker removed it :wink: )

Note: The main source files, where also non-mechanical changes had been done are: `file.F90`, `timearray.F90`, `initprogram.F90`, `fileacces.F90`, `inputdata.F90`, `timeprop.F90`

* [x] Testing on buildbot (success)

### CODE REVIEW

**Comment by aradi** [comment]
OK, don't merge yet as Intel MPI produces something very strange during `MPI_Finalize`, but only for the `legacy/file_access` test case. I'll have to investigate...

**Comment by aradi** [comment]
@thomas-niehaus Can you check, whether the realization as done here is sufficient for your integration efforts?

**Comment by thomas-niehaus** [comment]
If the default is stream, why do I get still the ASCII XplusY.DAT? In other words, you did not touch the excited state routines?

**Comment by aradi** [comment]
@thomas-niehaus This PR ensures only, that all binary I/O is made as stream, unless explicitly overridden by the user. I have converted all file I/O to the new type, but I did not change the format (binary/text) of any files yet. If an additional (or replacement) `XplusY.bin` file should be generated, this should be done in a new PR, once this machinery had been merged.

**Review by terminationshock — COMMENTED** [review]
Thanks for your work. I only have minor questions and a little request for a unit test.

File: src/dftbp/common/timerarray.F90
Comment (aradi) [reply src/dftbp/common/timerarray.F90:213]: It is an entry private to the module. (I tend to add the `_` suffix to private module/derived type variables/functions (similar to the `_` prefix convention in Python). One can then immediately see, that the entry is private. (Additionally, it also indicate that that entry might not have to be documented as detailed as a public one usually has to be.)

File: src/dftbp/common/file.F90
Comment (aradi) [reply src/dftbp/common/file.F90:318]: For files, opened via `%connectToFile()` or `openFile()` they are equivalent. However, one can also wrap a unit, with `%connectToUnit()` which had been opened already otherwise. We use this on the Fortran side of the C-API, where depending on the input, we either open a file ourselves, or use `output_unit` which is already open by default. (See https://github.com/aradi/dftbplus/blob/4ab9896810788816ab318f621f66ab8a8ef2d552/src/dftbp/api/mm/capi.F90#L517-L532)

**Review by terminationshock — APPROVED** [review]
Thank you for the test :+1:

**Pull request merged**
